cesium/internal/unary/delete.go

```diff
@@ -11,122 +11,85 @@ package unary
+// concurrently with other GarbageCollect methods.
+func (db *DB) GarbageCollect(ctx context.Context) error {
+	return db.wrapError(db.Domain.GarbageCollect(ctx))
+}
 
+func (db *DB) delete(ctx context.Context, tr telem.TimeRange) error {
+	if !tr.Valid() {
+		return errors.Newf("delete start %d cannot be after delete end %d", tr.Start, tr.End)
+	}
+
+	g, _, err := db.Controller.OpenAbsoluteGateIfUncontrolled(
+		tr,
+		control.Subject{Key: uuid.NewString(), Name: "delete_writer"},
+		func() (controlledWriter, error) {
+			return controlledWriter{Writer: nil, channelKey: db.Channel.Key}, nil
+		})
 	if err != nil {
 		return err
 	}
 
-	_, ok := g.Authorize()
-	if !ok {
-		g.Release()
-		return controller.Unauthorized(g.Subject.Name, db.Channel.Key)
-	}
+	g.Authorize()
+	defer g.Release()
 
-	i := db.Domain.NewIterator(domain.IteratorConfig{Bounds: telem.TimeRangeMax})
-	if ok := i.SeekGE(ctx, tr.Start); !ok {
-		return errors2.CombineErrors(i.Close(), errors.New("[cesium] Deletion Start TS not found"))
-	}
+	return db.Domain.Delete(ctx, db.calculateOffset, tr, db.Channel.DataType.Density())
+}
 
-	if i.TimeRange().Start.After(tr.Start) {
-		startOffset = 0
-		tr.Start = i.TimeRange().Start
-	} else {
-		approxDist, err := db.index().Distance(ctx, telem.TimeRange{
-			Start: i.TimeRange().Start,
-			End:   tr.Start,
-		}, false)
-		if err != nil {
-			return err
-		}
-		startOffset = approxDist.Upper
+// calculateOffset calculates the distance from a domain's start to the given time stamp.
+// Additionally, it "snaps" the timestamp to the nearest sample, depending on whether
+// this calculation is used to calculate the startOffset or endOffset.
+//
+// THIS METHOD SHOULD NOT BE CALLED BY UNARY! It should only be passed as a callback
+// to Domain.Delete.
+func (db *DB) calculateOffset(
+	ctx context.Context,
+	domainStart telem.TimeStamp,
+	ts *telem.TimeStamp,
+	isStart bool,
+) (int64, error) {
+	approxDist, err := db.index().Distance(ctx, telem.TimeRange{Start: domainStart, End: *ts}, true)
+	if err != nil {
+		return 0, err
 	}
-
-	startPosition := i.Position()
-
-	if ok := i.SeekLE(ctx, tr.End); !ok {
-		return errors2.CombineErrors(i.Close(), errors.New("[cesium] Deletion End TS not found"))
+	var (
+		offset   = approxDist.Upper
+		stampEnd = offset
+	)
+	if isStart {
+		// Note that stampEnd >= 0 since offset must be >= 1, as this function is
+		// only called if domainStart != ts and therefore approxDist.Upper must be
+		// at least 1.
+		stampEnd = offset - 1
 	}
-
-	if i.TimeRange().End.Before(tr.End) {
-		tr.End = i.TimeRange().End
-		endOffset = -1
-	} else {
-		approxDist, err := db.index().Distance(ctx, telem.TimeRange{
-			Start: i.TimeRange().Start,
-			End:   tr.End,
-		}, false)
+	if !approxDist.Exact() {
+		approxStamp, err := db.index().Stamp(ctx, domainStart, stampEnd, true)
 		if err != nil {
-			return errors2.CombineErrors(i.Close(), err)
+			return offset, err
+		}
+		if !approxStamp.Exact() {
```

Comment:
It seems like all of this logic is just useful when `isStart` is true. Why don't we pass two different callbacks to the domain `delete` method? One for resolving the start and one for the end. It would be more expressive.